source/common/upstream/upstream_impl.h

```diff
@@ -474,11 +488,24 @@ class HostImpl : public HostDescriptionImpl,
         host->handle_count_--;
       }
     }
-    const std::weak_ptr<const HostImpl> parent_;
+    const std::weak_ptr<const HostImplBase> parent_;
   };
   mutable std::atomic<uint32_t> handle_count_{};
 };
 
+class HostImpl : public HostImplBase, public HostDescriptionImpl {
+public:
+  HostImpl(ClusterInfoConstSharedPtr cluster, const std::string& hostname,
+           Network::Address::InstanceConstSharedPtr address, MetadataConstSharedPtr metadata,
+           uint32_t initial_weight, const envoy::config::core::v3::Locality& locality,
+           const envoy::config::endpoint::v3::Endpoint::HealthCheckConfig& health_check_config,
+           uint32_t priority, const envoy::config::core::v3::HealthStatus health_status,
+           TimeSource& time_source, const AddressVector& address_list = {})
```

Comment:
Yeah, right now, for several host types, the address_list is never populated and used, like static_cluster, strict_dns_cluster, original_dns_cluster,  ...

Maybe we can even further clean up the address_list from the HostDescriptionImpl, only keep address_ there?

Reply:
I don't think so, no. I think the impl I have is tuned for that is we don't populate even an empty vector or shared_ptr; it just defaults to nullptr (see upstream_impl.cc:418).